arch/src/aarch64/fdt.rs

```diff
@@ -0,0 +1,424 @@
+    //terminator in the reservemap list and whatever happened to be at the
+    //start of the FDT data section would end up being interpreted as
+    //reservemap entries.
+    // Safe since we previously allocated this array.
+    fdt_ret = unsafe { fdt_finish_reservemap(fdt.as_mut_ptr() as *mut c_void) };
+    if fdt_ret != 0 {
+        return Err(Error::FinishFDTReserveMap(io::Error::last_os_error()));
+    }
+
+    // For an explanation why these nodes were introduced in the blob take a look at
+    // https://github.com/torvalds/linux/blob/master/Documentation/devicetree/booting-without-of.txt#L845
+    // Look for "Required nodes and properties".
+
+    // Header or the root node as per above mentioned documentation.
+    begin_node(&mut fdt, "")?;
+    property_string(&mut fdt, "compatible", "linux,dummy-virt")?;
+    // For info on #address-cells and size-cells read "Note about cells and address representation"
+    // from the above mentioned txt file.
+    property_u32(&mut fdt, "#address-cells", ADDRESS_CELLS)?;
+    property_u32(&mut fdt, "#size-cells", SIZE_CELLS)?;
+    // This is not mandatory but we use it to point the root node to the node
+    // containing description of the interrupt controller for this VM.
+    property_u32(&mut fdt, "interrupt-parent", GIC_PHANDLE)?;
+    create_cpu_nodes(&mut fdt, num_cpus)?;
+    create_memory_node(&mut fdt, guest_mem)?;
+    create_chosen_node(&mut fdt, cmdline)?;
+    create_gic_node(&mut fdt, u64::from(num_cpus))?;
+    create_timer_node(&mut fdt)?;
+    create_psci_node(&mut fdt)?;
+
+    // End Header node.
+    end_node(&mut fdt)?;
+
+    let fdt_address = GuestAddress(get_fdt_addr(&guest_mem));
+    // Safe since we allocated fdt_final and previously passed in it's size
+    fdt_ret = unsafe { fdt_finish(fdt.as_mut_ptr() as *mut c_void) };
+    if fdt_ret != 0 {
+        return Err(Error::FinishFDTReserveMap(io::Error::last_os_error()));
+    }
+
+    // Allocate another buffer so we can format and then write fdt to guest.
+    let mut fdt_final = vec![0; FDT_MAX_SIZE];
+
+    // Safe because we allocated both arrays with the correct size
+    fdt_ret = unsafe {
+        fdt_open_into(
+            fdt.as_mut_ptr() as *mut c_void,
+            fdt_final.as_mut_ptr() as *mut c_void,
+            FDT_MAX_SIZE as i32,
+        )
+    };
+    if fdt_ret != 0 {
+        return Err(Error::FinishFDTReserveMap(io::Error::last_os_error()));
+    }
+
+    // Safe since we allocated fdt_final
+    fdt_ret = unsafe { fdt_pack(fdt_final.as_mut_ptr() as *mut c_void) };
+    if fdt_ret != 0 {
+        return Err(Error::FinishFDTReserveMap(io::Error::last_os_error()));
+    }
+    let written = guest_mem
+        .write_slice_at_addr(fdt_final.as_slice(), fdt_address)
+        .map_err(Error::WriteFDTToMemory)?;
+    if written < FDT_MAX_SIZE {
+        return Err(Error::IncompleteFDTMemoryWrite);
+    }
+    Ok(fdt_final)
+}
+
+// Following are auxiliary functions for appending nodes to FDT.
```

Comment:
Since this functions append nodes, would it make sense to prefix the names with `append_` ? For example: `append_begin_node` , `append_property_u32`

Otherwise it's hard to understand what `property_u32` does just by looking at its name.

Reply:
Done.